vowpalwabbit/vwdll.cpp

```diff
@@ -77,19 +77,30 @@ VW_DLL_MEMBER void      VW_CALLING_CONV VW_Finish(VW_HANDLE handle)
   VW::finish(*pointer);
 }
 
-VW_DLL_MEMBER VW_EXAMPLE VW_CALLING_CONV VW_ImportExample(VW_HANDLE handle, const char * label, VW_FEATURE_SPACE* features, size_t len)
+VW_DLL_MEMBER VW_EXAMPLE VW_CALLING_CONV VW_ImportExample(VW_HANDLE handle, const char * label, VW_FEATURE_SPACE features, size_t len)
 { vw * pointer = static_cast<vw*>(handle);
   VW::primitive_feature_space * f = reinterpret_cast<VW::primitive_feature_space*>( features );
   return static_cast<VW_EXAMPLE>(VW::import_example(*pointer, label, f, len));
 }
 
+VW_DLL_MEMBER VW_FEATURE_SPACE VW_CALLING_CONV VW_InitializeFeatureSpaces(size_t len)
+{
```

Comment:
Why are these needed precisely?

Reply:
In other words, what is the missing functionality amongst existing interfaces?

Reply:
With existing API the only way to "construct" example - VwExample ImportExample(FeatureSpace, Label);
But the only way to get feature space - FeatureSpace ExportExample(VwExample).
So I do not see any way to construct example without introducing functions required for initializing, adding features / namespaces etc to feature space. Please correct me if I am wrong.

Reply:
I guess the question is: What is the api?  Is it vwdll.cpp?  Or is it vw.h?  To the extent that vwdll.cpp requires additional functions exposing vw.h that seems fine.  To the extent that vw.h is missing things, I'd like to understand more deeply.  Is vw.h missing things?  If not, can these extra interface functions just call the vw.h interface?

Reply:
My understanding that cpp api is vw.h while c one is vwdll.
vw.h is not missing anything - at least allows to create example, since it is exposing full primitive_feature_space implementation:
https://github.com/VowpalWabbit/vowpal_wabbit/blob/92064660a778d343285e92620fa8077d41db841c/vowpalwabbit/vw.h#L54
So given current vw.h code, direct manipulations with primitive_feature space is technically "using vw.h interface".
vw.h can be refactored to hide its feature_space implementation details behind the interfaces, but still looks like corresponding flat c methods have to be added to vwdll.

Reply:
Good.  So, can we make sure the added vwdll functions use functions in vw.h?  Currently, it seems they are not, so this is sort-of creating an alternate interface.

Reply:
Sure, let me try.

Reply:
Sent the update draft, please take a look.